proposals/p2240.md

```diff
@@ -0,0 +1,758 @@
+trade a readability problem for a teachability problem, by introducing more
+syntax that programmers need to learn before they can even read variadic
+function signatures.
+
+### Keyword syntax
+
+> **TODO:** Decide whether to make this the primary proposal, and restructure
+> the discussion as appropriate.
+
+We could build the syntax around keywords rather than symbolic tokens, much like
+the
+[Swift proposal](https://github.com/apple/swift-evolution/blob/main/proposals/0393-parameter-packs.md).
+For concreteness, the keywords I will discuss here are as follows:
+
+| Symbolic token | Keyword replacement               |
+| -------------- | --------------------------------- |
+| `[:]`          | `expand`                          |
+| `...,`         | `repeat` and `unpack` (see below) |
+| `...and`       | `all_of`                          |
+| `...or`        | `any_of`                          |
+| `...{`         | `repeat {`                        |
+
+This approach seems to work best in combination with some other nominally
+independent changes:
+
+-   Forbid `expand` in pattern contexts.
+-   Allow `repeat` only in expression contexts and `unpack` only in pattern
+    contexts.
+-   Require any usage of a pack-type variable to be prefixed with a new keyword
+    `each`.
+
+With these changes, the examples given earlier would read as follows:
+
+```carbon
+// Computes the sum of its arguments, which are i64s
+fn SumInts(unpack params: expand [i64;]) -> i64 {
+  var sum: i64 = 0;
+  repeat {
+    sum += each params;
+  }
+  return sum;
+}
+```
+
+```carbon
+// Concatenates its arguments, which are all convertible to String
+fn StrCat[T:! [ConvertibleToString;]](unpack params: expand T) -> String {
+  var len: i64 = 0;
+  repeat { len += (each params).Length(); }
+  var result: String = "";
+  result.Reserve(len);
+  repeat { result.Append((each params).ToString()); }
+  return result;
+}
+```
+
+```carbon
+// Returns the minimum of its arguments, which must all have the same type T.
+//
+// Note that this implementation is not recursive. We split the parameters into
+// first and rest in order to forbid calling `Min` with no arguments.
+fn Min[T:! Comparable & Value](first: T, unpack rest: expand [T;]) -> T {
+  var result: T = first;
+  repeat {
+    if (each rest < result) {
+      result = each rest;
+    }
+  }
+  return result;
+}
+```
+
+```carbon
+// Invokes f, with the tuple `args` as its arguments.
+fn Apply[T:! [type;], F:! CallableWith(repeat expand T)]
+        (f: F, args: (repeat expand T)) -> auto {
+  return f(repeat expand args);
+}
+```
+
+```carbon
+// Takes an arbitrary number of vectors with arbitrary element types, and
+// returns a vector of tuples where the i'th element of the vector is
+// a tuple of the i'th elements of the input vectors.
+fn Zip[ElementTypes:! [type;]]
+      (unpack vectors: Vector(expand ElementTypes))
+      -> Vector((repeat expand ElementTypes)) {
```

Comment:
```suggestion
fn Zip[pack ElementTypes:! type]
      (pack vectors: Vector(each ElementTypes))
      -> Vector((repeat each ElementTypes)) {
```

Reply:
Calling @zygoloid , who was probably not notified of the updates to this thread.

* Re: "if I want to know the type of a name, looking to the right of the `:` will give me the wrong answer", the thing to the _left_ of the `:!` is `pack T` not `T`, so it seems reasonable that the thing to the right of the `:!` is not the type of `T`.
* Re: `pack` in the deduced parameter list -- the deduced parameter list is _part_ of the pattern if I'm not mistaken. It also has the "arity determined by the scrutinee" property.

Reply:
Also, the intent of this suggestion is to be consistent with the behavior of expansions where non-packs values are repeated for each expansion.

Reply:
The more recent `each` suggestion recovers the generality here by separating the information that a binding is a pack from the position where it's expanded -- `(...(each x: i32, each y: i32))` would be a valid pattern, matching a list of pairs of `i32`s.

I think that approach also addresses your concern about the thing on the right of the `:` not necessarily being the type of the thing on the left -- when we declare `each x: T`, the type of `each x` is `T`. (@chandlerc described this as being somewhat like a name prefix / sigil, like `$` / `@` in perl.)

Regarding implicit parameter lists -- yeah, they're not really lists of general patterns, and they don't really have scrutinees (except for `self`). But I think it's important that moving a parameter between an implicit and explicit parameter list uses the same syntax. I think it would be surprising if you can directly introduce packs in pattern syntax, but you have to introduce an array when writing something analogous in an implicit parameter list. For example, I think it's valuable that these two functions use the same syntax for their packs of types:

```
// Make a tuple containing unformed values.
fn MakeUnformedTuple(...each T:! type) -> (...each T);

// Make a tuple containing specific values.
fn MakeTuple[...each T:! type](x: T) -> (...each T);
```